src/crimson/osd/pg_backend.cc

```diff
@@ -1163,7 +1163,88 @@ PGBackend::omap_get_keys(
       ll_read_errorator::pass_further{}
     );
 }
+static
+int do_omap_val_cmp(
+  std::map<std::string, ceph::bufferlist, std::less<>> out,
+  std::map<std::string, std::pair<bufferlist, int>> assertions)
+{
+  int r = 0;
+  bufferlist empty;
+  for (std::map<std::string, std::pair<bufferlist, int> >::iterator
+    i = assertions.begin();
+    i != assertions.end();
+    ++i) {
+    auto out_entry = out.find(i->first);
+    bufferlist &bl = (out_entry != out.end()) ?
+      out_entry->second : empty;
+    switch (i->second.second) {
+      case CEPH_OSD_CMPXATTR_OP_EQ:
+        if (!(bl == i->second.first)) {
+          r = -ECANCELED;
```

Comment:
I think you can replace the `if (r < 0) break;` with `returns`. In turn, this would allow to errorate the entire method.